### ISSUE TITLE
failure activating enviroment when submitted to nodes

### ISSUE THREAD

**Issue by Yue-Jiang** [conda/conda#3001]
When submitting jobs on a HPC cluster, from time to time i run into such errors

```
FileNotFoundError: [Errno 2] No such file or directory: '/somepath/env/conda3/bin/conda'
```

when trying to submit jobs like this in the pipeline:

```
source /somepath/env/lib/miniconda3/bin/activate /somepath/env/conda3 && do_stuff
```

But the `'/somepath/env/conda3/bin/conda'` is definitely there. A simple rerun will usually fix it so It's been hard to reproduce this behavior. The guess is that it's caused by multiple jobs running on different nodes are trying to source the same environment?

**Comment by dkoppstein** [comment]
Indeed, I am getting the same error. To clarify, here's the full traceback:

```
Traceback (most recent call last):
  File "/large/dkoppstein/test_160329_1/rules/../env/lib/miniconda3/bin/conda", line 6, in <module>
    sys.exit(main())
  File "/large/dkoppstein/test_160329_1/env/lib/miniconda3/lib/python3.5/site-packages/conda/cli/main.py", line 48, in main
    activate.main()
  File "/large/dkoppstein/test_160329_1/env/lib/miniconda3/lib/python3.5/site-packages/conda/cli/activate.py", line 137, in main
    conda.install.symlink_conda(prefix, root_dir, shell)
  File "/large/dkoppstein/test_160329_1/env/lib/miniconda3/lib/python3.5/site-packages/conda/install.py", line 589, in symlink_conda
    symlink_conda_hlp(prefix, root_dir, where, symlink_fn)
  File "/large/dkoppstein/test_160329_1/env/lib/miniconda3/lib/python3.5/site-packages/conda/install.py", line 603, in symlink_conda_hlp
    os.remove(prefix_file)
FileNotFoundError: [Errno 2] No such file or directory: '/large/dkoppstein/test_160329_1/env/conda2/bin/conda'
```

I think it's caused by the call to `os.remove` in `symlink_conda_hlp`, seems like the file might transiently not exist, so if jobs on other nodes are running `source activate` then this step will fail? 

Currently testing whether rolling back to `4.0.10` works as was suggested [here](https://github.com/conda/conda/issues/2837).

**Comment by dkoppstein** [comment]
Indeed, rolling back to 4.0.10 fixes the issue. Thus, I think this is a bug in the current master.

**Comment by andreas-wilm** [comment]
Same problem here, even though only parallel submission of >>100 jobs seems to trigger the problem.

Things work so far after downgrading to 4.0.10.

**Comment by tomkinsc** [comment]
We're seeing the same issue, and it does appear to be resolved by downgrading to 4.0.10.

**Comment by andreas-wilm** [comment]
I recognize most of the names in this thread...Can't be something specific to snakemake right?

**Comment by dkoppstein** [comment]
I doubt it, I think it just is a side effect of submitting hundreds of jobs at a time that all call `source activate` to a conda installation on NFS. It just so happens that we're the ones doing that the most ;)

**Comment by dkoppstein** [comment]
Conda fix appears to be in progress. 
conda/conda#3210

**Comment by tomkinsc** [comment]
Has there been any progress on this?

**Comment by andreas-wilm** [comment]
Bump

**Comment by tomkinsc** [comment]
Ok, I took a look at the conda source. It looks like an NFS delay-related race condition, where one cluster job tries to create a symlink after a symlink from a second cluster job has appeared between the existence check and symlink creation in the first job. The function responsible was [more basic](https://github.com/conda/conda/blob/4.0.10/conda/install.py#L414) in conda v4.0.10  than the symlink function [in v4.2.9](https://github.com/conda/conda/blob/4.2.9/conda/install.py#L535) (it didn't try to remove the symlink if it already exists), but that's not the real issue here. The exception handler in the 4.2.9 version that that would ordinarily ignore such errors is susceptible to the same race condition.  (`os.path.lexists(prefix_file)` may not be True in the exception handler, so the exception gets raised and the activate/deactivate script dies)

If we can patch [`conda/install.py`](https://github.com/conda/conda/blob/master/conda/install.py#L174) to something like the following I believe it should hopefully resolve the issue:

``` python
def symlink_conda_hlp(prefix, root_dir, where, symlink_fn):
    scripts = ["conda", "activate", "deactivate"]
    prefix_where = join(prefix, where)
    if not isdir(prefix_where):
        os.makedirs(prefix_where)
    for f in scripts:
        root_file = join(root_dir, where, f)
        prefix_file = join(prefix_where, f)
        try:
            # try to kill stale links if they exist
            if os.path.lexists(prefix_file):
                rm_rf(prefix_file)
            # if they're in use, they won't be killed.  Skip making new symlink.
            if not os.path.lexists(prefix_file):
                symlink_fn(root_file, prefix_file)
        except (IOError, OSError) as e:
            if (os.path.lexists(prefix_file) and
                    (e.errno in (errno.EPERM, errno.EACCES, errno.EROFS, errno.EEXIST))):
                log.debug("Cannot symlink {0} to {1}. Ignoring since link already exists."
                          .format(root_file, prefix_file))
            elif (e.errno == errno.ENOENT):
                log.debug("Problem with symlink management {0} {1}. File may have been removed by another concurrent process." .format(root_file, prefix_file))
            elif (e.errno == errno.EEXIST)):
                log.debug("Problem with symlink management {0} {1}. File may have been created by another concurrent process." .format(root_file, prefix_file))
            else:
                raise
```

With this change conda should chug along if it can't create the symlink during activation, and it should still work since _one_ of the cluster jobs will have created it. There's a chance there'll be a new race condition if something tries to use the link before NFS can propagate it, but hopefully that will be much less likely.

Thoughts? Anyone care to try this out on their system with a newer version of conda? If it is working for others I can open a PR.

**Comment by kalefranz** [comment]
Fixed with #4345 and #4346
